projects/igniteui-angular/src/lib/services/exporter-common/base-export-service.ts

```diff
@@ -318,15 +325,174 @@ export abstract class IgxBaseExporter {
+
+        for(const entry of records) {
+            const expansionStateVal = grid.expansionStates.has(entry) ? grid.expansionStates.get(entry) : false;
+
+            const dataWithoutChildren = Object.keys(entry)
+                .filter(k => columnFields.includes(k))
+                .reduce((obj, key) => {
+                    obj[key] = entry[key];
+                    return obj;
+                }, {});
+
+            const hierarchicalGridRecord: IExportRecord = {
+                data: dataWithoutChildren,
+                level: 0,
+                type: ExportRecordType.HierarchicalGridRecord,
+                owner: grid,
+            };
+
+            this.flatRecords.push(hierarchicalGridRecord);
+
+            for (const key of childLayoutKeys) {
+                const island = grid.childLayoutList.filter(l => l.key === key)[0];
+                const islandGrid = childGrids.filter(g => g.key === island.key && g.data === entry[key])[0];
+
+                const keyRecordData = this.prepareIslandData(islandGrid, entry[key]) || [];
+
+                this.getAllChildColumnsAndData(island, keyRecordData, expansionStateVal, childGrids, grid);
+            }
+        }
+    }
+
+    private prepareIslandData(islandGrid: IgxHierarchicalGridComponent, data: any): any {
+        if (islandGrid !== undefined) {
+            const hasFiltering = (islandGrid.filteringExpressionsTree &&
+                    islandGrid.filteringExpressionsTree.filteringOperands.length > 0) ||
+                (islandGrid.advancedFilteringExpressionsTree &&
+                    islandGrid.advancedFilteringExpressionsTree.filteringOperands.length > 0);
+
+            const hasSorting = islandGrid.sortingExpressions &&
+                islandGrid.sortingExpressions.length > 0;
+
+            const skipOperations =
+                (!hasFiltering || !this.options.ignoreFiltering) &&
+                (!hasSorting || !this.options.ignoreSorting);
+
+                if (skipOperations) {
+                    data = islandGrid.filteredSortedData;
+                } else {
+                    if (hasFiltering && !this.options.ignoreFiltering) {
+                        const filteringState: IFilteringState = {
+                            expressionsTree: islandGrid.filteringExpressionsTree,
+                            advancedExpressionsTree: islandGrid.advancedFilteringExpressionsTree,
+                        };
+
+                        filteringState.strategy = islandGrid.filterStrategy;
+
+                        data = DataUtil.filter(data, filteringState, islandGrid);
+                    }
+
+                    if (hasSorting && !this.options.ignoreSorting) {
+                        this._sort = cloneValue(islandGrid.sortingExpressions[0]);
+
+                        data = DataUtil.treeGridSort(data, islandGrid.sortingExpressions, islandGrid.sortStrategy);
+                    }
+                }
+        }
+
+        return data;
```

Comment:
If there is not `islandGrid`, we should apply the filtering and sorting of the island directly on the data.

Reply:
This is still not resolved.